pkg/protocol/engine/tipmanager/v1/tip_metadata.go

```diff
@@ -15,341 +13,211 @@ type TipMetadata struct {
+	return t.isWeakTip
 }
 
-// deriveIsWeaklyReferencedByTips derives the isReferencedByOtherTips property.
-func (t *TipMetadata) deriveIsWeaklyReferencedByTips() {
-	t.connectedWeakChildren.OnUpdate(func(_, connectedWeakChildren int) {
-		t.isReferencedByOtherTips.Compute(func(_ bool) bool {
-			return connectedWeakChildren > 0 || t.isStronglyReferencedByOtherTips.Get()
-		})
-	})
-
-	t.isStronglyReferencedByOtherTips.OnUpdate(func(_, isStronglyReferencedByOtherTips bool) {
-		t.isReferencedByOtherTips.Compute(func(_ bool) bool {
-			return isStronglyReferencedByOtherTips || t.connectedWeakChildren.Get() > 0
-		})
-	})
+// IsOrphaned returns a ReadableVariable that indicates if the block was orphaned.
+func (t *TipMetadata) IsOrphaned() reactive.ReadableVariable[bool] {
+	return t.isOrphaned
 }
 
-// deriveIsStrongTip derives the isStrongTip property.
-func (t *TipMetadata) deriveIsStrongTip() {
-	t.isStronglyReferencedByOtherTips.OnUpdate(func(_, isStronglyReferencedByOtherTips bool) {
-		t.isStrongTip.Compute(func(_ bool) bool {
-			return !isStronglyReferencedByOtherTips && t.isStrongTipPoolMember.Get()
-		})
-	})
-
-	t.isStrongTipPoolMember.OnUpdate(func(_, isStrongTipPoolMember bool) {
-		t.isStrongTip.Compute(func(_ bool) bool {
-			return isStrongTipPoolMember && !t.isStronglyReferencedByOtherTips.Get()
-		})
-	})
+// Evicted exposes an event that is triggered when the block is evicted.
+func (t *TipMetadata) Evicted() reactive.Event {
+	return t.evicted
 }
 
-// deriveIsWeakTip derives the isWeakTip property.
-func (t *TipMetadata) deriveIsWeakTip() {
-	t.isReferencedByOtherTips.OnUpdate(func(_, isReferencedByOtherTips bool) {
-		t.isWeakTip.Compute(func(_ bool) bool {
-			return !isReferencedByOtherTips && t.isWeakTipPoolMember.Get()
-		})
-	})
-
-	t.isWeakTipPoolMember.OnUpdate(func(_, isWeakTipPoolMember bool) {
-		t.isWeakTip.Compute(func(_ bool) bool {
-			return isWeakTipPoolMember && !t.isReferencedByOtherTips.Get()
-		})
-	})
-}
+// connectStrongParent sets up the parent and children related properties for a strong parent.
+func (t *TipMetadata) connectStrongParent(strongParent *TipMetadata) {
+	t.stronglyOrphanedStrongParents.Monitor(strongParent.isStronglyOrphaned)
 
-// deriveIsOrphaned derives the isOrphaned property.
-func (t *TipMetadata) deriveIsOrphaned() {
-	t.isMarkedOrphaned.OnUpdate(func(_, isMarkedOrphaned bool) {
-		t.isOrphaned.Compute(func(_ bool) bool {
-			return isMarkedOrphaned || t.orphanedStrongParents.Get() > 0
-		})
-	})
-
-	t.orphanedStrongParents.OnUpdate(func(_, orphanedStrongParents int) {
-		t.isOrphaned.Compute(func(_ bool) bool {
-			return orphanedStrongParents > 0 || t.isMarkedOrphaned.Get()
-		})
-	})
+	// unsubscribe when the parent is evicted, since we otherwise continue to hold a reference to it.
+	unsubscribe := strongParent.stronglyConnectedStrongChildren.Monitor(t.isStronglyConnectedToTips)
+	strongParent.evicted.OnUpdate(func(_, _ bool) { unsubscribe() })
```

Comment:
would this also trigger the unsubscribe if the strongParent is already evicted in the moment we call `OnUpdate` ?
Is this a possible memory leak?

Reply:
Yes. If `evicted` has been triggered before this line, then the callback is executed immediately, so unsubscribe will be called in the described case.

Reply:
ok cool, then this is resolved! :+1: